ops/relation.py

```diff
@@ -0,0 +1,548 @@
+            pairs that are all the acceptable provider
+            specifications. The keys in these specifications are the
+            resource names strings. And the values are version
+            specification strings. The version specification strings
+            can by in any form that is compatible with the
+            [semver](https://pypi.org/project/semver/) Python
+            package. A valid example of the `consumes` dictionary is
+            Example::
+
+            {'consumes': {'mysql': '>5.0.2', 'mariadb': '<=6.1.0'}}
+        multi: a boolean flag that indicates if the :class:`Consumer` object
+            should allow multiple relations with the same relation name. By
+            default this is `False`.
+
+    """
+    on = ConsumerEvents()
+    stored = StoredState()
+
+    def __init__(self, charm, name, consumes, multi=False):
+        super().__init__(charm, name)
+
+        self.name = name
+        self.consumes = consumes
+        self.multi_mode = multi
+        self.stored.set_default(relation_id=None)
+
+        events = charm.on[name]
+        self.framework.observe(events.relation_changed, self.on_provider_changed)
+        self.framework.observe(events.relation_broken, self.on_provider_broken)
+        self.framework.observe(charm.on.upgrade_charm, self.validate_provider)
+
+    def on_provider_changed(self, event):
+        """Validate provider on relation changed event
+
+        This method checks the provider for compatibility with the
+        consumer every time a relation changed event is raised. The
+        provider is also checked to ensure it is ready to service
+        requests. In response to these checks any of the following
+        events may be raised.
+
+        - :class:`ProviderAvailable`
+        - :class:`ProviderInvalid`
+        - :class:`ProviderUnready`
+
+        Args:
+            event: an event object. It is expected that the event object
+                contains a key `provider_data` whose value is all the data
+                forwarded by the :class:`Provider` object.
+        """
+        rdata = event.relation.data[event.app]
+        logger.debug("Got data from provider : {}".format(rdata))
+        provider_data = rdata.get('provider_data')
+        consumed = self.consumes
+        if provider_data:
+            data = json.loads(provider_data)
+            try:
+                provides = data['provides']
+            except KeyError:
+                # provider has not set any specification
+                # so no compatibility checks are done
+                # and no events are raised
+                logger.warning('Provider not specified')
+                return
+        else:
+            logger.debug('No provider data')
+            # provider has not given any information
+            # so provider will not be made available (as yet)
+            return
+
+        ready = data.get('ready')
+        if not ready:
+            self.on.unready.emit()
```

Comment:
This feels a little strange to emit Unready multiple times, if there are provider changes that still result in unready (providers may have reason to change content in their provider_data that don't strictly result in the next event being ready).
I don't know if this matters, or if you just need to document that you may get multiple Unready events.

Reply:
You make a good point. However kindly note this question relates to another discussion regarding "ready" and "unready" being set by the provider charm, say during an upgrade. Hence the need to let the consumer charm know of every provider change if the status is unready. An alternative is to keep track of the number of times unready message has been sent to the consumer charm. I fear such an implementation can have unusual edge cases that are difficult to handle. But if this is strongly preferred I am willing to give it another thought after some exploration here. 

Also if the unready event were forwarding data then the it would certainly be useful for the consumer charm to progress itself incrementally as and when the provider charm updates data incrementally. I would be interested hear if such behaviour has known or anticipated use cases. If so I can implement the data forwarding in the unready event.

Also I will certainly add additional documentation pertaining to multiple unready event, if so decided.

Thoughts ?

ATTN: @jlounder will be grateful for any insights you may have here ?

Reply:
Added clarification to docstring stating the event may be raised multiple times.